### ISSUE TITLE
Intermitent java.util.NoSuchElementException: 2 when parsing script

### ISSUE THREAD

**Issue by JLLeitschuh** [shyiko/ktlint#287]
I'm not 100% certain if this is a Ktlint issue or not, but I'm hoping that if it's not a ktlint issue, perhaps you'd be able to point me in the right direction for solving this issue.

This is the contents of `Application.gradle.kts`:
```kotlin
description = "Sub-Projects are the various independent Applications that make up Fitting."

subprojects {
    afterEvaluate {
        dependencies {
            "implementation"(group = "com.github.ajalt", name = "clikt", version = "1.3.0")
            "runtimeOnly"(project(":Logging:Log4Production"))
        }
    }
}

```
Every few builds we get this strange error:
```
> Task :Application:spotlessKotlinGradle FAILED

Step 'ktlint' found problem in 'Application/Application.gradle.kts':

2

java.util.NoSuchElementException: 2
	at kotlin.jvm.internal.ArrayIterator.next(ArrayIterator.kt:11)
	at kotlin.sequences.FlatteningSequence$iterator$1.ensureItemIterator(Sequences.kt:253)
	at kotlin.sequences.FlatteningSequence$iterator$1.hasNext(Sequences.kt:242)
	at org.jetbrains.kotlin.script.CachingSequence$CachingIterator.hasNext(KotlinScriptDefinitionProvider.kt:102)
	at org.jetbrains.kotlin.script.LazyScriptDefinitionProvider.findScriptDefinition(KotlinScriptDefinitionProvider.kt:115)
	at org.jetbrains.kotlin.script.LazyScriptDefinitionProvider.isScript(KotlinScriptDefinitionProvider.kt:84)
	at org.jetbrains.kotlin.parsing.KotlinParser.parse(KotlinParser.java:46)
	at org.jetbrains.kotlin.psi.stubs.elements.KtFileElementType.doParseContents(KtFileElementType.java:82)
	at org.jetbrains.kotlin.com.intellij.psi.tree.IFileElementType.parseContents(IFileElementType.java:49)
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.LazyParseableElement.lambda$ensureParsed$0(LazyParseableElement.java:193)
	at org.jetbrains.kotlin.com.intellij.psi.impl.DebugUtil.performPsiModification(DebugUtil.java:544)
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.LazyParseableElement.ensureParsed(LazyParseableElement.java:192)
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.LazyParseableElement.getFirstChildNode(LazyParseableElement.java:247)
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.LazyParseableElement.getFirstChildNode(LazyParseableElement.java:40)
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.RecursiveTreeElementWalkingVisitor$ASTTreeGuide.getFirstChild(RecursiveTreeElementWalkingVisitor.java:48)
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.RecursiveTreeElementWalkingVisitor$ASTTreeGuide.getFirstChild(RecursiveTreeElementWalkingVisitor.java:35)
	at org.jetbrains.kotlin.com.intellij.util.WalkingState.next(WalkingState.java:73)
	at org.jetbrains.kotlin.com.intellij.util.WalkingState.walkChildren(WalkingState.java:61)
	at org.jetbrains.kotlin.com.intellij.util.WalkingState.elementStarted(WalkingState.java:52)
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.RecursiveTreeElementWalkingVisitor.visitNode(RecursiveTreeElementWalkingVisitor.java:86)
	at org.jetbrains.kotlin.com.intellij.psi.impl.GeneratedMarkerVisitor.visitNode(GeneratedMarkerVisitor.java:28)
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.RecursiveTreeElementWalkingVisitor.visitComposite(RecursiveTreeElementWalkingVisitor.java:81)
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.CompositeElement.acceptTree(CompositeElement.java:131)
	at org.jetbrains.kotlin.com.intellij.psi.impl.PsiFileFactoryImpl.markGenerated(PsiFileFactoryImpl.java:228)
	at org.jetbrains.kotlin.com.intellij.psi.impl.PsiFileFactoryImpl.trySetupPsiForFile(PsiFileFactoryImpl.java:130)
	at org.jetbrains.kotlin.com.intellij.psi.impl.PsiFileFactoryImpl.createFileFromText(PsiFileFactoryImpl.java:88)
	at org.jetbrains.kotlin.com.intellij.psi.impl.PsiFileFactoryImpl.createFileFromText(PsiFileFactoryImpl.java:73)
	at org.jetbrains.kotlin.com.intellij.psi.impl.PsiFileFactoryImpl.createFileFromText(PsiFileFactoryImpl.java:67)
	at org.jetbrains.kotlin.com.intellij.psi.impl.PsiFileFactoryImpl.createFileFromText(PsiFileFactoryImpl.java:61)
	at com.github.shyiko.ktlint.core.KtLint.format(KtLint.kt:320)
	at com.github.shyiko.ktlint.core.KtLint.formatScript(KtLint.kt:298)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.diffplug.spotless.kotlin.KtLintStep$State.lambda$createFormat$1(KtLintStep.java:112)
	at com.diffplug.spotless.FormatterStepImpl$Standard.format(FormatterStepImpl.java:78)
	at com.diffplug.spotless.FormatterStep$Strict.format(FormatterStep.java:76)
	at com.diffplug.spotless.Formatter.compute(Formatter.java:230)
	at com.diffplug.spotless.Formatter.isClean(Formatter.java:167)
	at com.diffplug.gradle.spotless.SpotlessTask.check(SpotlessTask.java:263)
	at com.diffplug.gradle.spotless.SpotlessTask.performAction(SpotlessTask.java:205)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
	at org.gradle.api.internal.project.taskfactory.IncrementalTaskAction.doExecute(IncrementalTaskAction.java:50)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:131)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:300)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:292)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:174)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:120)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:99)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:77)
	at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
	at org.gradle.api.internal.tasks.execution.SkipCachedTaskExecuter.execute(SkipCachedTaskExecuter.java:105)
	at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ResolveBuildCacheKeyExecuter.execute(ResolveBuildCacheKeyExecuter.java:79)
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
	at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
	at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.run(EventFiringTaskExecuter.java:51)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:300)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:292)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:174)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:46)
	at org.gradle.execution.taskgraph.LocalTaskInfoExecutor.execute(LocalTaskInfoExecutor.java:42)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareWorkItemExecutor.execute(DefaultTaskExecutionGraph.java:277)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareWorkItemExecutor.execute(DefaultTaskExecutionGraph.java:262)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:135)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:130)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.execute(DefaultTaskPlanExecutor.java:200)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.executeWithWork(DefaultTaskPlanExecutor.java:191)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.run(DefaultTaskPlanExecutor.java:130)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
```

System info:

Ktlint version `0.27.0`.
```
------------------------------------------------------------
Gradle 4.10.1
------------------------------------------------------------

Build time:   2018-09-12 11:33:27 UTC
Revision:     76c9179ea9bddc32810f9125ad97c3315c544919

Kotlin DSL:   1.0-rc-6
Kotlin:       1.2.61
Groovy:       2.4.15
Ant:          Apache Ant(TM) version 1.9.11 compiled on March 23 2018
JVM:          1.8.0_181 (Oracle Corporation 25.181-b13)
OS:           Linux 3.10.0-862.el7.x86_64 amd64

```

**Comment by shyiko** [comment]
Hey, Jonathan. 

Looks like kotlin compiler was unable to parse whatever `Application/Application.gradle.kts` contained at the time. 

```sh
$ cat Application.gradle.kts 
description = "Sub-Projects are the various independent Applications that make up Fitting."

subprojects {
    afterEvaluate {
        dependencies {
            "implementation"(group = "com.github.ajalt", name = "clikt", version = "1.3.0")
            "runtimeOnly"(project(":Logging:Log4Production"))
        }
    }
}

$ ktlint --format Application.gradle.kts
...
[DEBUG] 786ms / 1 file(s) / 0 error(s)
```

Any chance you still have that Application/Application.gradle.kts? (Application.gradle.kts provided above appears to be valid)

**Comment by JLLeitschuh** [comment]
Hey @shyiko,
This is the file as is when this failed. This is a very strange issue that only occurs on our Jenkins build slave. None of the developers are able to reproduce the issue locally.

Curiosity, does the Kotlin `PsiFileFactoryImpl` call out to the Kotlin Compiler Daemon?

**Comment by shyiko** [comment]
> Curiosity, does the Kotlin PsiFileFactoryImpl (or anything it calls) call out to the Kotlin Compiler Daemon?

doesn't look like it

Try adding
```
ktlint --format Application/Application.gradle.kts || cat -A Application/Application.gradle.kts
```
If it doesn't fail then the first thing I'd do would be to check the dependecy graph (of `:Application:spotlessKotlinGradle`). Make sure only one version of kotlin runtime is present.

**Comment by JLLeitschuh** [comment]
What version of kotlin is `0.27.0` supposed to rely upon. I was reading the POM file but the only version it has specified in the published pom is `${kotlin.version}`.

I was reading this file here:
https://search.maven.org/artifact/com.github.shyiko/ktlint/0.27.0/jar

**Comment by shyiko** [comment]
@JLLeitschuh 
https://search.maven.org/artifact/com.github.shyiko.ktlint/pom/0.27.0/pom
`<kotlin.version>1.2.51</kotlin.version>`

**Comment by eskatos** [comment]
We observe the same intermitent issue when running ktlint using ktlint-gradle on a Gradle build with multiple projects run with Gradle parallel execution enabled where several ktlint tasks are running concurrently.

**Comment by JLLeitschuh** [comment]
Oh, it's not just us & it's not just spotless!! This is really good to know!! Thanks, @eskatos

**Comment by JLLeitschuh** [comment]
I was able to reliably reproduce this issue on my local machine in rare cases.
I have a 70+ build script project.
I ran the following command in a loop until it failed (it took several iterations to finally fail).

```
./gradlew clean spotlessKotlinGradleCheck --parallel --max-workers=32 --stacktrace
```

```
------------------------------------------------------------
Gradle 4.10.1
------------------------------------------------------------

Build time:   2018-09-12 11:33:27 UTC
Revision:     76c9179ea9bddc32810f9125ad97c3315c544919

Kotlin DSL:   1.0-rc-6
Kotlin:       1.2.61
Groovy:       2.4.15
Ant:          Apache Ant(TM) version 1.9.11 compiled on March 23 2018
JVM:          1.8.0_191 (Oracle Corporation 25.191-b12)
OS:           Mac OS X 10.13.6 x86_64

```

**Comment by JLLeitschuh** [comment]
I'm suspecting that there might be an issue with the logic here:
https://github.com/JetBrains/kotlin/blob/d3f52936473a9391f1d8902d699b8df531f5d8cf/compiler/psi/src/org/jetbrains/kotlin/script/KotlinScriptDefinitionProvider.kt#L52-L90

In particular, I think the underlying assumption made by this code is that Sequences can be assumed to be thread-safe from a reader perspective, and I don't think that's the case.

**Comment by shyiko** [comment]
Thank you for getting to the bottom of it, @JLLeitschuh!

![](https://images-na.ssl-images-amazon.com/images/I/512OOMv4ZEL.jpg)

**Comment by shyiko** [comment]
I believe this one is fixed in ktolin-compiler@1.3.20 (bundled with ktlint@0.30.0).
